File: android/measure/src/main/java/sh/measure/android/tracing/InternalSpan.kt
package sh.measure.android.tracing

import sh.measure.android.Measure

/**
 * Implements a [Span] and adds some internal functions to it.
 */
internal interface InternalSpan : Span {
    /**
     * Gets the name identifying this span.
     *
     * @return The name assigned to this span when it was created.
     */
    val name: String

    /**
     * Gets the session identifier associated with this span. A v4-UUID string.
     *
     * @return The unique identifier for the session this span belongs to
     */
    val sessionId: String

    /**
     * Gets the timestamp when this span was started.
     *
     * @return The start time in milliseconds since epoch, obtained via [Measure.getCurrentTime].
     */
    val startTime: Long

    /**
     * Gets the list of time-based checkpoints added to this span.
     *
     * @return A mutable list of [Checkpoint] objects, each representing a significant
     * point in time during the span's lifecycle
     *
     * Note: Checkpoints can be added during the span's lifetime using [setCheckpoint] to mark
     * important events or transitions within the traced operation.
     */
    val checkpoints: MutableList<Checkpoint>

    /**
     * Gets the map of attributes attached to this span.
     *
     * @return The attributes added to the span.
     */
    val attributes: Map<String, Any?>

    /**
     * Gets the current status of this span, indicating its outcome or error state.
     *
     * @return [SpanStatus] The status of the span.
     */
    fun getStatus(): SpanStatus

    /**
     * Returns a modifiable map of attributes.
     */
    fun getAttributesMap(): MutableMap<String, Any?>

    /**
     * Adds an attribute.
     */
    fun setAttribute(attribute: Pair<String, Any?>)

    /**
     * Converts the span to a data class for further processing and export.
     */
    fun toSpanData(): SpanData
}

File: flutter/packages/measure_flutter/lib/src/tracing/internal_span.dart
import 'package:measure_flutter/src/tracing/span.dart';
import 'package:measure_flutter/src/tracing/span_data.dart';
import 'package:measure_flutter/src/tracing/span_status.dart';

import 'checkpoint.dart';

// Implements a [Span] and adds some internal functions to it.
abstract class InternalSpan extends Span {
  /// Gets the name identifying this span.
  ///
  /// Returns the name assigned to this span when it was created.
  String get name;

  /// Gets the timestamp when this span was started.
  ///
  /// Returns the start time in milliseconds since epoch, obtained via [Measure.getCurrentTime].
  int get startTime;

  /// Gets the list of time-based checkpoints added to this span.
  ///
  /// Returns a mutable list of [Checkpoint] objects, each representing a significant
  /// point in time during the span's lifecycle
  ///
  /// Note: Checkpoints can be added during the span's lifetime using [setCheckpoint] to mark
  /// important events or transitions within the traced operation.
  List<Checkpoint> get checkpoints;

  /// Gets the map of attributes attached to this span.
  ///
  /// Returns the attributes added to the span.
  Map<String, dynamic> get attributes;

  /// Gets the current status of this span, indicating its outcome or error state.
  ///
  /// Returns [SpanStatus] The status of the span.2
  SpanStatus getStatus();

  /// Returns a modifiable map of attributes.
  Map<String, dynamic> getAttributesMap();

  /// Returns a modifiable map of user-defined attributes.
  Map<String, dynamic> getUserDefinedAttrs();

  /// Adds an attribute to this span.
  void setInternalAttribute(MapEntry<String, dynamic> attribute);

  /// Updates the sampling status of the span.
  ///
  /// @param sampled whether the span is sampled or not.
  void setSamplingRate(bool sampled);

  /// Converts the span to a data class for further processing and export.
  SpanData toSpanData();
}
